posts/2023-06-01-Rust-1.70.0.md

```diff
@@ -0,0 +1,74 @@
+
+```rust
+use std::sync::OnceLock;
+
+fn main() {
+    let winner_lock = OnceLock::new();
+
+    let winner = std::thread::scope(|s| {
+        s.spawn(|| winner_lock.set("thread"));
+
+        std::thread::yield_now(); // give them a chance...
+
+        winner_lock.get_or_init(|| "main")
+    });
+
+    println!("{winner} wins!");
+}
+```
+
+Crates such as `lazy_static` and `once_cell` have filled this need in the past, but now these building blocks are part of the standard library, ported from `once_cell`'s `unsync` and `sync` modules. There are still more methods that may be stabilized in the future, as well as companion `LazyCell` and `LazyLock` types that store their initializing function, but this first step in stabilization should already cover many use cases.
+
+### Named levels of debug information
+
+The `-Cdebuginfo` compiler option has previously only supported numbers 0..=2 for increasing amounts of debugging information, where Cargo defaults to 2 in dev and test profiles and 0 in release and bench profiles. These debug levels can now be set by name: "none" (0), "limited" (1), and "full" (2), as well as two new levels, "line-directives-only" and "line-tables-only".
+
+The Cargo and rustc documentation both called level 1 "line tables only" before, but it was more than that with information about all functions, just not types and variables. That level is now called "limited", and the new "line-tables-only" level is further reduced to the minimum needed for backtraces with filenames and line numbers. This may eventually become the level used for `-Cdebuginfo=1`. The other `line-directives-only` level is intended for NVPTX profiling, and is otherwise not recommended.
+
+### Enforced stability in the `test` CLI
+
+When `#[test]` functions are compiled, the executable gets a command-line interface from the `test` crate. This CLI has a number of options, including some that are not yet stabilized and require specifying `-Zunstable-options` as well, like many other commands in the Rust toolchain. However, while that's only intended to be allowed in nightly builds, that restriction wasn't active in `test` -- until now. Starting with 1.70.0, stable and beta builds of Rust will no longer allow unstable `test` options, making them truly nightly-only as documented.
+
+### Stabilized APIs
+
+- [`thing`](link)
+
+These APIs are now stable in const contexts:
+
+- [`const_thing`](link)
+
+### Other changes
+
+Check out everything that changed in [Rust](https://github.com/rust-lang/rust/blob/stable/RELEASES.md#version-1700-2023-06-01), [Cargo](https://github.com/rust-lang/cargo/blob/master/CHANGELOG.md#cargo-1700-2023-06-01, and [Clippy](https://github.com/rust-lang/rust-clippy/blob/master/CHANGELOG.md#rust-170).
```

Comment:
I'm a little concerned about linking to `RELEASES.md` in this release. The file is very frequently no longer rendering. I suspect this will result in quite a few reports and complaints. I don't know if there are any good options here. One idea is to link to the raw page instead of the rendered one (https://raw.githubusercontent.com/rust-lang/rust/stable/RELEASES.md). Another option is to do a quick hack to cut out a bunch of content (move it to another file?).  I'm also not sure if anyone has reached out to GitHub?

I'm also still willing to help with coming up with some other solution, but it would be helpful to find some way to make a decision on which direction to go.

Reply:
> Another option is to do a quick hack to cut out a bunch of content (move it to another file?). I'm also not sure if anyone has reached out to GitHub?

I think reducing the file size is your best option!

`RELEASES.md` is pushing 632 kilobytes in size. That's pretty massive in terms of rendering HTML out of markdown on the fly (the HTML ends up much larger than the markdown). Not to mention the typical person is only looking at recent release info. However, this is my _personal_ opinion and **_not_** an official response from GitHub. 😉 (If it were up to me, I'd explore rewriting the rendering in Rust 😆).

Reply:
Ok, I've set the links to the tag.